scipy/special/xsf/dual.h

```diff
@@ -103,10 +120,9 @@ class dual<T, Order> {
 
     dual &operator/=(const dual &other) {
         for (size_t i = 0; i <= Order; ++i) {
-            T coef = 1; // binomial coefficient
+            // General Leibniz Rule
             for (size_t j = 1; j <= i; ++j) {
-                data[i] -= coef * other.data[j] * data[i - j];
-                coef *= T(i - j) / T(j + 1);
+                data[i] -= detail::fast_binom<T>(i, j) * other.data[j] * data[i - j];
```

Comment:
Here it looks like the loop is `for j in range(1, i+1):`

But this doesn't pass:

```python3
from scipy.special import binom

i = 10
coef = 1
for j in range(1, i+1):
    assert coef == binom(i, j)
    coef *= (i - j) / (j + 1)
```
What did I miss?

Reply:
On close inspection, I think that the formula for dividing dual numbers given here is not correct, either before or after this change. The reason this didn't get caught in #21483 is because there are no instances where nontrivial dual numbers actual get divided. The correct formula can be found in the answers to the [mathstackexchange question](https://math.stackexchange.com/q/5357/7323). I blindly substituted in `fast_binom` in this PR, but the `coef` being computed in the loop originally is `binom(i, j) / i` not `binom(i, j)`. I think neither result is correct though.

Reply:
OK, I think I'll go ahead and merge this since it's not related to the PR. It appears that the actual formula will involve a binomial coefficient, so it will be useful to have this `fast_binom` when it's time to correct this and add a test that exercises it.

Reply:
@steppi @mdhaber I don't think it's as simple as that, and I think what I had originally was right. I think it may have been in fact `binom(i - 1, k) = binom(i, k) / i`. Can you please revert this until either of you can come up with a counterexample? This may have introduced a bug.

Reply:
No need to revert because there is an issue against this - gh-21665. We know that `binom(i - 1, k) = binom(i, k) / i`, but @steppi didn't think that was right, either, so I didn't ask for it to be changed here. If it really should be `binom(i, k) / i`, then it will be simple to fix in a separate PR that also adds a test.

Reply:
@izaid sent me the reference, https://dspace.mit.edu/bitstream/handle/1721.1/151501/tan-songchen-csesm-ccse-2023-thesis%20Songchen%20Tan.pdf?sequence=1&isAllowed=y, with formula

![image](https://github.com/user-attachments/assets/f3e73c3e-caf4-4b36-959a-c733ffa70098)

which is close to what he has, but if I'm reading it correctly, when his original code had this

```C++
T coef = 1; // binomial coefficient
            for (size_t j = 1; j <= i; ++j) {
                data[i] -= coef * other.data[j] * data[i - j];
                coef *= T(i - j) / T(j + 1);
            }
```

it should have been

```
T coef = 1; // binomial coefficient
            for (size_t j = 1; j <= i; ++j) {
                data[i] -= coef * data[j] * other.data[i - j];
                coef *= T(i - j) / T(j + 1);
            }
```

Reply:
I think `other.data` contains `f` actually. Because we are composing functions as we go, so we already had `g(x)`.